src/libstore/path.hh

```diff
@@ -1,59 +1,78 @@
 #pragma once
 
-#include "rust-ffi.hh"
+#include "types.hh"
 
 namespace nix {
 
-/* See path.rs. */
-struct StorePath;
-
 class Store;
+struct Hash;
 
-extern "C" {
-    void ffi_StorePath_drop(void *);
-    bool ffi_StorePath_less_than(const StorePath & a, const StorePath & b);
-    bool ffi_StorePath_eq(const StorePath & a, const StorePath & b);
-    unsigned char * ffi_StorePath_hash_data(const StorePath & p);
-}
-
-struct StorePath : rust::Value<3 * sizeof(void *) + 24, ffi_StorePath_drop>
+class StorePath
 {
+    std::string baseName;
+
+    StorePath(const StorePath & path)
+        : baseName(path.baseName)
+    { }
+
+public:
+
+    /* Size of the hash part of store paths, in base-32 characters. */
+    constexpr static size_t HashLen = 32; // i.e. 160 bits
+
     StorePath() = delete;
 
-    static StorePath make(std::string_view path, std::string_view storeDir);
+    StorePath(std::string_view baseName);
 
-    static StorePath make(unsigned char hash[20], std::string_view name);
+    StorePath(const Hash & hash, std::string_view name);
 
-    static StorePath fromBaseName(std::string_view baseName);
+    StorePath(StorePath && path)
+        : baseName(std::move(path.baseName))
+    { }
 
-    rust::String to_string() const;
+    StorePath & operator = (StorePath && path)
+    {
+        baseName = std::move(path.baseName);
+        return *this;
+    }
+
+    std::string_view to_string() const
+    {
+        return baseName;
+    }
 
     bool operator < (const StorePath & other) const
     {
-        return ffi_StorePath_less_than(*this, other);
+        return baseName < other.baseName;
     }
 
     bool operator == (const StorePath & other) const
     {
-        return ffi_StorePath_eq(*this, other);
+        return baseName == other.baseName;
     }
 
     bool operator != (const StorePath & other) const
     {
-        return !(*this == other);
+        return baseName != other.baseName;
     }
 
-    StorePath clone() const;
+    StorePath clone() const
+    {
+        return StorePath(*this);
+    }
```

Comment:
I hope we can get rid of this and go back to just using idiomatic copy constructors, and STL copying in particular?

Reply:
Yes, but I didn't want to clutter this PR to much. Will do that afterwards.

Reply:
Sorry, I meant to say "after" and didn't. Great plan, happy to do some of that for you if you like :).